sched/event/event_post.c

```diff
@@ -0,0 +1,155 @@
+ * Name: nxevent_post_internal
+ ****************************************************************************/
+
+static inline_function int nxevent_post_internal(FAR sem_t *sem)
+{
+  int semcount;
+
+  nxsem_get_value(sem, &semcount);
+  if (semcount < 1)
+    {
+      return nxsem_post(sem);
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxevent_post
+ *
+ * Description:
+ *   Post one or more events to an event object.
+ *
+ *   This routine posts one or more events to an event object. All tasks
+ *   waiting on the event object event whose waiting conditions become
+ *   met by this posting immediately unpend.
+ *
+ *   Posting differs from setting in that posted events are merged together
+ *   with the current set of events tracked by the event object.
+ *
+ * Input Parameters:
+ *   event  - Address of the event object
+ *   events - Set of events to post to event
+ *          - Set events to 0 will be considered as any,
+ *            waking up the waiting thread immediately.
+ *
+ * Returned Value:
+ *   This is an internal OS interface and should not be used by applications.
+ *   It follows the NuttX internal error return policy:  Zero (OK) is
+ *   returned on success.  A negated errno value is returned on failure.
+ *
+ * Assumptions:
+ *   This function may be called from an interrupt handler.
+ *
+ ****************************************************************************/
+
+int nxevent_post(FAR nxevent_t *event, nxevent_mask_t events)
+{
+  FAR nxevent_wait_t *wait;
+  FAR nxevent_wait_t *tmp;
+  nxevent_mask_t revents;
+  irqstate_t flags;
+  int ret = 0;
+
+  if (event == NULL)
+    {
+      return -EINVAL;
+    }
+
+  flags = enter_critical_section();
+
+  if (!list_is_empty(&event->list))
+    {
+      /* Hold schedule lock here to avoid context switch if post high
+       * priority task.
+       */
+
+      sched_lock();
+
+      list_for_every_entry_safe(&event->list, wait, tmp,
+                                nxevent_wait_t, node)
+        {
+          if (events == 0)
+            {
+              revents = wait->expect ? wait->expect : ~0;
+            }
+          else
+            {
+              revents = events;
+            }
+
+          /* Specified events */
+
+          if (wait->expect)
+            {
+              event->events |= (revents & wait->expect);
+              if ((wait->expect & event->events) == wait->expect)
+                {
+                  ret = nxevent_post_internal(&wait->sem);
+                }
+            }
+
+          /* Any events */
+
+          else
+            {
+              wait->expect |= revents;
+              ret = nxevent_post_internal(&wait->sem);
+            }
+        }
+
```

Comment:
need remove the flags which already post to the waiter

Reply:
Done